File: generate_resume.py
import os
import sys
import json
import subprocess

from dotenv import load_dotenv
import os

load_dotenv()  # Load environment variables from .env file

AUTH_TOKEN = os.getenv("AUTH_TOKEN")  # Get the token from the environment

OUTPUT_DIR = "generated_resumes"

def latex_escape(text):
    if not text:
        return ""
    return str(text).replace("&", "\\&").replace("%", "\\%").replace("_", "\\_")\
        .replace("#", "\\#").replace("{", "\\{").replace("}", "\\}").replace("^", "\\^{}")

def format_skills(skills):
    if not skills:
        return ""
    lines = []
    if "TechnicalSkills" in skills and skills["TechnicalSkills"]:
        lines.append("\\textbf{Technical Skills:} " + latex_escape(", ".join(skills["TechnicalSkills"])))
    if "SoftSkills" in skills and skills["SoftSkills"]:
        lines.append("\\textbf{Soft Skills:} " + latex_escape(", ".join(skills["SoftSkills"])))
    return "\\section*{Skills}\n" + "\n".join(lines)

def format_education(education):
    if not education:
        return ""
    blocks = []
    for edu in education:
        block = []
        if edu.get("Degree"):
            block.append(f"\\textbf{{{latex_escape(edu['Degree'])}}}")
        if edu.get("Institution"):
            block.append(latex_escape(edu["Institution"]))
        if edu.get("Year"):
            block.append(f"({latex_escape(edu['Year'])})")
        if edu.get("CGPA"):
            block.append(f"\\textit{{CGPA:}} {latex_escape(edu['CGPA'])}")
        if edu.get("Percentage"):
            block.append(f"\\textit{{Percentage:}} {latex_escape(edu['Percentage'])}")
        if edu.get("Achievements"):
            for ach in edu["Achievements"]:
                block.append(f"\\quad - {latex_escape(ach)}")
        if edu.get("RelevantCoursework"):
            block.append("\\textbf{Relevant Coursework:} " + ", ".join([latex_escape(c) for c in edu["RelevantCoursework"]]))
        blocks.append(" \\\\ ".join(block))
    return "\\section*{Education}\n" + "\n\\vspace{0.1cm}\n".join(blocks)

def format_experience(experience):
    if not experience:
        return ""
    blocks = []
    for exp in experience:
        block = []
        if exp.get("Title"):
            block.append(f"\\textbf{{{latex_escape(exp['Title'])}}}")
        if exp.get("Company"):
            block.append(latex_escape(exp["Company"]))
        if exp.get("Duration"):
            block.append(f"\\hfill {latex_escape(exp['Duration'])}")
        if exp.get("Description"):
            block.append(latex_escape(exp["Description"]))
        blocks.append(" \\\\ ".join(block))
    return "\\section*{Experience}\n" + "\n\\vspace{0.1cm}\n".join(blocks)

def format_projects(projects):
    if not projects:
        return ""
    blocks = []
    for proj in projects:
        block = []
        if proj.get("Name"):
            block.append(f"\\textbf{{{latex_escape(proj['Name'])}}}")
        if proj.get("Description"):
            block.append(latex_escape(proj["Description"]))
        if proj.get("Technologies"):
            block.append(f"\\textit{{Technologies:}} {latex_escape(proj['Technologies'])}")
        blocks.append(" \\\\ ".join(block))
    return "\\section*{Projects}\n" + "\n\\vspace{0.1cm}\n".join(blocks)

def format_certifications(certifications):
    if not certifications:
        return ""
    blocks = []
    for cert in certifications:
        block = []
        if cert.get("Name"):
            block.append(f"\\textbf{{{latex_escape(cert['Name'])}}}")
        if cert.get("Issuer"):
            block.append(latex_escape(cert["Issuer"]))
        if cert.get("Year"):
            block.append(f"({latex_escape(cert['Year'])})")
        blocks.append(" — ".join(block))
    return "\\section*{Certifications}\n" + "\n\\vspace{0.1cm}\n".join(blocks)

def format_links(links):
    if not links:
        return ""
    return " \\\\ ".join([f"\\href{{{link}}}{{{link}}}" for link in links])

def format_summary(summary):
    if not summary:
        return ""
    return f"\\section*{{Summary}}\n{latex_escape(summary)}"

def generate_resume_pdf(data):
    if not os.path.exists(OUTPUT_DIR):
        os.makedirs(OUTPUT_DIR)

    user_id = data.get("userId", "user")
    job_id = data.get("job", {}).get("id", "job")
    pdf_basename = f"{user_id}_{job_id}"
    pdf_path = os.path.join(OUTPUT_DIR, f"{pdf_basename}.pdf")
    tex_path = os.path.join(OUTPUT_DIR, f"{pdf_basename}.tex")

    # Build contact information
    contact_lines = []
    if data.get("Name"):
        contact_lines.append(f"{{\\LARGE \\textbf{{{latex_escape(data['Name'])}}}}}")
    if data.get("Email"):
        contact_lines.append(latex_escape(data["Email"]))
    if data.get("Phone"):
        contact_lines.append(latex_escape(data["Phone"]))
    if data.get("Links"):
        contact_lines.append(format_links(data["Links"]))

    # Build LaTeX document
    latex = f"""
\\documentclass[10pt]{{article}}
\\usepackage[margin=0.5in]{{geometry}}
\\usepackage{{hyperref}}
\\usepackage{{parskip}}
\\pagenumbering{{gobble}}

\\begin{{document}}

\\begin{{center}}
{chr(10).join(contact_lines)}
\\end{{center}}

{format_summary(data.get("Summary", ""))}

{format_skills(data.get("Skills", {}))}

{format_education(data.get("Education", []))}

{format_experience(data.get("Experience", []))}

{format_projects(data.get("Projects", []))}

{format_certifications(data.get("Certifications", []))}

\\end{{document}}
"""

    # Write LaTeX file
    with open(tex_path, "w") as f:
        f.write(latex)

    try:
        # Run pdflatex twice to resolve references
        for _ in range(2):
            result = subprocess.run([
                "pdflatex",
                "-interaction=nonstopmode",
                f"-jobname={pdf_basename}",
                "-output-directory", OUTPUT_DIR,
                tex_path
            ], check=True, capture_output=True, text=True)

        # Clean up auxiliary files
        aux_files = ['.aux', '.log', '.out']
        for ext in aux_files:
            aux_path = os.path.join(OUTPUT_DIR, f"{pdf_basename}{ext}")
            if os.path.exists(aux_path):
                os.remove(aux_path)

        return {
            "user": user_id,
            "job": job_id,
            "message": "Resume generated",
            "pdf": pdf_path
        }

    except subprocess.CalledProcessError as e:
        print("❌ Error generating PDF:", e)
        print("❌ stdout:", e.stdout)
        print("❌ stderr:", e.stderr)
        return None

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python3 generate_resume.py <data_file>")
        sys.exit(1)

    with open(sys.argv[1], 'r') as f:
        data = json.load(f)
    
    result = generate_resume_pdf(data)
    if result:
        print(result)

File: generateResume.js
const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');
const { execAsync } = require('./utils');

async function generateResume(user, job) {
    try {
        const data = {
            user: user,
            job: job
        };

        const latexContent = generateLatex(data);
        const timestamp = Date.now();
        const sanitizedUserName = user.Name.replace(/[^a-zA-Z0-9]/g, '_');
        const sanitizedJobTitle = (job.title || 'Unknown_Job').replace(/[^a-zA-Z0-9]/g, '_');
        const pdfPath = `generated_resumes/${sanitizedUserName}_${sanitizedJobTitle}_${timestamp}.pdf`;

        await generatePDF(latexContent, pdfPath);

        return pdfPath;
    } catch (error) {
        console.error('Error generating resume:', error);
        throw error;
    }
}

function sanitizeFilename(filename) {
    // Replace special characters with underscores
    return filename.replace(/[^a-zA-Z0-9]/g, '_');
}

async function generatePDF(user, job) {
    try {
        // Sanitize the filename
        const sanitizedJobTitle = sanitizeFilename(job.title);
        const outputFilename = `${user.name.replace(/\s+/g, '_')}_${sanitizedJobTitle}`;
        const outputPath = path.join('generated_resumes', outputFilename);
        
        // Create LaTeX content
        const latexContent = generateLaTeX(user, job);
        
        // Write LaTeX file
        fs.writeFileSync(`${outputPath}.tex`, latexContent);
        
        // Generate PDF
        const result = await execAsync(`pdflatex -interaction=nonstopmode -jobname="${outputFilename}" -output-directory generated_resumes "${outputPath}.tex"`);
        
        // Clean up auxiliary files
        try {
            fs.unlinkSync(`${outputPath}.aux`);
            fs.unlinkSync(`${outputPath}.log`);
            fs.unlinkSync(`${outputPath}.out`);
        } catch (err) {
            console.warn('Warning: Could not clean up auxiliary files:', err.message);
        }
        
        return `${outputFilename}.pdf`;
    } catch (error) {
        console.error('Error generating PDF:', error);
        throw error;
    }
}

function generateLatex(data) {
    const { user, job } = data;
    
    // Format education entries
    const educationEntries = user.Education.map(edu => {
        let entry = `\\textbf{${edu.Year || ''}} & \\textbf{${edu.Degree || ''}}, \\textit{${edu.Institution || ''}} \\\\\n`;
        if (edu.Grade) entry += `& \\textbf{${edu.Grade}} \\\\\n`;
        if (edu.Coursework) entry += `& ${edu.Coursework} \\\\[3pt]\n`;
        if (edu.Achievements && Array.isArray(edu.Achievements)) {
            entry += `& \\begin{itemize}\n`;
            edu.Achievements.forEach(achievement => {
                entry += `    \\item ${achievement}\n`;
            });
            entry += `\\end{itemize}\n`;
        }
        return entry;
    }).join('\n');

    // Format skills
    const technicalSkills = user.Skills.TechnicalSkills ? user.Skills.TechnicalSkills.join(', ') : '';
    const softSkills = user.Skills.SoftSkills ? user.Skills.SoftSkills.join(', ') : '';

    // Format projects
    const projectEntries = user.Projects.map(project => {
        let entry = `\\textbf{${project.Name}} – \\textit{${project.Year || ''}}\n\n`;
        entry += `\\vspace{2pt}\n\\begin{itemize}\n`;
        entry += `  \\item ${project.Description}\n`;
        if (project.Technologies) {
            entry += `  \\item Technologies used: ${project.Technologies}\n`;
        }
        entry += `\\end{itemize}\n\n\\vspace{4pt}\n`;
        return entry;
    }).join('\n');

    // Format experience
    const experienceEntries = user.Experience.map(exp => {
        let entry = `\\textbf{${exp.Title}} at \\textit{${exp.Company}} – \\textit{${exp.Duration}}\n\n`;
        entry += `\\vspace{2pt}\n\\begin{itemize}\n`;
        if (exp.Description) {
            entry += `  \\item ${exp.Description}\n`;
        }
        if (exp.Responsibilities && Array.isArray(exp.Responsibilities)) {
            exp.Responsibilities.forEach(resp => {
                entry += `  \\item ${resp}\n`;
            });
        }
        entry += `\\end{itemize}\n\n\\vspace{4pt}\n`;
        return entry;
    }).join('\n');

    // Format certifications
    const certificationEntries = user.Certifications.map(cert => {
        return `\\textbf{${cert.Name}} from \\textit{${cert.Issuer}} – \\textit{${cert.Year}}\n\n`;
    }).join('\n');

    // Generate the LaTeX document
    return `\\documentclass[10pt]{article}
\\usepackage[a4paper,margin=1.2cm]{geometry}
\\usepackage{enumitem}
\\usepackage{xcolor}
\\usepackage{titlesec}
\\usepackage{tabularx}
\\usepackage[hidelinks]{hyperref}
\\usepackage{multicol}
\\setlist[itemize]{left=0pt,label=--,itemsep=2pt,topsep=0pt}
\\renewcommand{\\familydefault}{\\sfdefault}

% Section formatting
\\titleformat{\\section}{\\color{blue}\\Large\\bfseries}{}{0em}{}[\\titlerule]

\\begin{document}

\\begin{center}
    {\\LARGE \\textbf{${user.Name}}} \\\\
    \\vspace{2pt}
    \\href{mailto:${user.Email}}{${user.Email}} \\quad | \\quad ${user.Phone} \\quad | \\quad \\href{${user.Links[0] || '#'}}{${user.Links[0] || 'LinkedIn'}}
\\end{center}

\\vspace{4pt}
\\section*{Professional Summary}
${user.Summary}

\\vspace{4pt}
\\section*{Educational Qualification}
\\begin{tabularx}{\\linewidth}{@{}lX@{}}
${educationEntries}
\\end{tabularx}

\\vspace{4pt}
\\section*{Skills}

\\textbf{Technical Skills:} \\\\
${technicalSkills}

\\vspace{6pt}
\\textbf{Soft Skills:} \\\\
${softSkills}

\\vspace{4pt}
\\section*{Experience}

${experienceEntries}

\\vspace{4pt}
\\section*{Projects}

${projectEntries}

\\vspace{4pt}
\\section*{Certifications}

${certificationEntries}

\\end{document}`;
}

module.exports = generateResume; 